src/ripple/rpc/handlers/LedgerData.cpp

```diff
@@ -78,28 +79,29 @@ Json::Value doLedgerData (RPC::Context& context)
     jvResult[jss::ledger_hash] = to_string (lpLedger->info().hash);
     jvResult[jss::ledger_index] = std::to_string(lpLedger->info().seq);
 
-    Json::Value& nodes = (jvResult[jss::state] = Json::arrayValue);
+    Json::Value& nodes = jvResult[jss::state];
 
-    for(auto const& sle : lpLedger->sles)
+    while ((key = lpLedger->succ(*key)))
     {
-       if (limit-- <= 0)
-       {
-           auto marker = sle->key();
-           jvResult[jss::marker] = to_string (--marker);
-           break;
-       }
-
-       if (isBinary)
-       {
-           Json::Value& entry = nodes.append (Json::objectValue);
-           entry[jss::data] = serializeHex(*sle);
-           entry[jss::index] = to_string (sle->key());
-       }
-       else
-       {
-           Json::Value& entry = nodes.append (sle->getJson (0));
-           entry[jss::index] = to_string (sle->key());
-       }
+        auto sle = lpLedger->read(keylet::unchecked(*key));
+        if (limit-- <= 0)
+        {
```

Comment:
The `marker` is a SHAMap key, and when a request includes one, it will find the very next ledger entry in the map with a key that compares greater. It's used to pick things up from a prior request where there was more items than the limit. But if all items have been seen and the limit not reached there's no need for a marker.

You can see that the `key` variable is set the marker, and if one isn't present, then it will be set to 0 (all indexes, being high entropy hashes, will compare greater than uin256(0))